modules/lang-painless/src/main/java/org/elasticsearch/painless/Definition.java

```diff
@@ -698,22 +736,41 @@ private void addStruct(ClassLoader whitelistClassLoader, Whitelist.Struct whitel
                 javaClass = Class.forName(whitelistStruct.javaClassName, true, whitelistClassLoader);
             } catch (ClassNotFoundException cnfe) {
                 throw new IllegalArgumentException("invalid java class name [" + whitelistStruct.javaClassName + "]" +
-                        " for struct [" + whitelistStruct.painlessTypeName + "]");
+                        " for struct [" + painlessTypeName + "]");
             }
         }
 
-        Struct existingStruct = structsMap.get(whitelistStruct.painlessTypeName);
+        Struct existingStruct = structsMap.get(painlessTypeName);
 
         if (existingStruct == null) {
-            Struct struct = new Struct(whitelistStruct.painlessTypeName, javaClass, org.objectweb.asm.Type.getType(javaClass));
+            Struct struct = new Struct(painlessTypeName, javaClass, org.objectweb.asm.Type.getType(javaClass));
+            structsMap.put(painlessTypeName, struct);
 
-            structsMap.put(whitelistStruct.painlessTypeName, struct);
-            simpleTypesMap.put(whitelistStruct.painlessTypeName, getTypeInternal(whitelistStruct.painlessTypeName));
+            if (simpleTypesMap.containsKey(painlessTypeName) == false && simpleTypesMap.containsKey(importedPainlessTypeName) == false) {
+                simpleTypesMap.put(painlessTypeName, getTypeInternal(painlessTypeName));
+            }
         } else if (existingStruct.clazz.equals(javaClass) == false) {
-            throw new IllegalArgumentException("struct [" + whitelistStruct.painlessTypeName + "] is used to " +
+            throw new IllegalArgumentException("struct [" + painlessTypeName + "] is used to " +
                     "illegally represent multiple java classes [" + whitelistStruct.javaClassName + "] and " +
                     "[" + existingStruct.clazz.getName() + "]");
         }
+
+        if (whitelistStruct.onlyFQNJavaClassName == false) {
+            existingStruct = structsMap.get(importedPainlessTypeName);
+
+            if (existingStruct == null) {
+                structsMap.put(importedPainlessTypeName, structsMap.get(painlessTypeName));
+                Type painlessType = simpleTypesMap.remove(painlessTypeName);
```

Comment:
I'm a little confused by this block. Why is the type removed only to be added back?

Reply:
For the doc generator to work simple types names must not contain multiple entries of the same type with multiple names, so this removes the fully-qualified name in favor the short name for now to be consistent with what already exists.  This definitely needs to be cleaned up but not in this PR.